### PR TITLE
MRG : API decoding

### PR DESCRIPTION
**Pull request by trachelr**
So I've created an class Linear classifier which has attributes filters_ (coef_)  and patterns_
There's also a method plot_pattern (I can also add a method plot_filters for consistency)

- [ ] Attributes for Scaler (mean_ and std_)
- [ ] ConcatenateChannels should be renamed as EpochsVectorizer
- [ ] FilterEstimator should call a private function in transform

### CODE REVIEW

File: examples/decoding/plot_decoding_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+# Testing will be done with a running classifier
+epochs = Epochs(raw, events, event_id, tmin, tmax, proj=True, picks=picks,
+                baseline=None, preload=True, add_eeg_ref=False)
+epochs_train = epochs.crop(tmin=1., tmax=2., copy=True)
+labels = epochs.events[:, -1] - 2
+
+# import a few transformer objects from mne.decoding
+from mne.decoding.transformer import Scaler, PSDEstimator, ConcatenateChannels
+# import a linear classifier from mne.decoding
+from mne.decoding.classifier import LinearClassifier
```
Comment (agramfort) [inline examples/decoding/plot_decoding_pipeline.py]: I would add stuff to init to be able to do:

from mne.decoding import Scaler, PSDEstimator, ConcatenateChannels, LinearClassifier

File: examples/decoding/plot_decoding_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
+import numpy as np
```
Comment (agramfort) [inline examples/decoding/plot_decoding_pipeline.py]: not doc header, no author and no licence.

File: examples/decoding/plot_decoding_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+                     ('cat', cat), ('linear', clf)))
+
+pipeline.fit(epochs_train.get_data(), labels)
+# get the patterns
+patterns = pipeline.steps[-1][1].patterns_
+
+# sampling of psd estimates at 1 Hz (1 sec epochs)
+info['sfreq'] = 1
+patterns = EvokedArray(patterns.reshape(info['nchan'], -1), info, tmin=fmin)
+patterns.plot_topomap(layout=layout, scale_time=1, time_format='%01d Hz')
```
Comment (agramfort) [inline examples/decoding/plot_decoding_pipeline.py]: time_format='%01d Hz' ?

File: examples/decoding/plot_decoding_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+                   bandwidth=6, adaptive=True)
+sc  = Scaler(info)
+cat = ConcatenateChannels()
+clf = LinearClassifier(LogisticRegression())
+psd = PSDEstimator(sfreq=info['sfreq'], 
+                   fmin=fmin, fmax=fmax, 
+                   bandwidth=3)
+
+pipeline = Pipeline((('psd', psd), ('scaler', sc),
+                     ('cat', cat), ('linear', clf)))
```
Comment (agramfort) [inline examples/decoding/plot_decoding_pipeline.py]: does it really make sense to use a log reg on psd to do decoding of motor imagery. To me not really. It's more natural to do it on sample and do it on evoked response.

File: mne/decoding/__init__.py [hunk 5]
```diff
@@ -1,5 +1,7 @@
-from .classifier import Scaler, FilterEstimator
-from .classifier import PSDEstimator, ConcatenateChannels
+from .transformer import Scaler, FilterEstimator
+from .transformer import PSDEstimator, ConcatenateChannels
+#from .transformer import compute_patterns
```
Comment (agramfort) [inline mne/decoding/__init__.py]: ?

File: mne/decoding/classifier.py [hunk 6]
```diff
@@ -1,455 +1,144 @@
...
-from ..filter import (low_pass_filter, high_pass_filter, band_pass_filter,
-                      band_stop_filter)
-from ..time_frequency import multitaper_psd
-from ..externals import six
-from ..utils import _check_type_picks
-
-
-class Scaler(TransformerMixin):
-    """Standardizes data across channels
+class LinearClassifier():
```
Comment (agramfort) [inline mne/decoding/classifier.py]: class LinearClassifier(object):

File: mne/decoding/regressor.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
+# Authors: Romain Trachel <trachelr@gmail.com>
+#          Alexandre Gramfort <alexandre.gramfort@telecom-paristech.fr>
+#
+# License: BSD (3-clause)
+
+import numpy as np
+
+
+class LinearRegressor():
```
Comment (agramfort) [inline mne/decoding/regressor.py]: object missing

File: mne/decoding/regressor.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    """
+    This object clones a Linear Model from sklearn
```
Comment (agramfort) [inline mne/decoding/regressor.py]: don't name the package sklearn but scikit-learn

File: mne/decoding/regressor.py [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+    patterns_ : ndarray
+        If fit, the patterns used to restore M/EEG signals, else None.
+    
+    References
+    ----------
+    """
+    def __init__(self, reg):
+        if reg is None:
+            from sklearn.linear_model import LinearRegression
+            reg = LinearRegression()
```
Comment (agramfort) [inline mne/decoding/regressor.py]: then make reg=None by default.

File: mne/decoding/regressor.py [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+        
+        References
+        ----------
+        """
+        if not isinstance(X, np.ndarray):
+            raise ValueError("X should be of type ndarray (got %s)."
+                             % type(X))
+        if not isinstance(y, np.ndarray):
+            raise ValueError("y should be of type ndarray (got %s)."
+                             % type(y))
```
Comment (agramfort) [inline mne/decoding/regressor.py]: y could be a list. In scikit-learn we support array-like and use the check function in validation.py

File: mne/decoding/regressor.py [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+        X = np.atleast_2d(X)
+        if len(X.shape) != 2:
+            raise ValueError("X dimension should be 2 (n_epochs x n_features)"
+                             " instead of ", X.shape)
+        # fit the regressor
+        self.reg.fit(X, y)
+        # computes the patterns
+        if hasattr(self.reg, 'coef_'):
+            self.patterns_ = np.dot(X.T, np.dot(X, self.reg.coef_.T))
+            self.filters_ = self.reg.coef_
```
Comment (agramfort) [inline mne/decoding/regressor.py]: raise an error if coef_ is not here. Then compute the patterns all the time.

File: mne/decoding/regressor.py [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+        y : array, shape=(n_epochs)
+            The class for each epoch.
+
+        Returns
+        -------
+        y_pred : array, shape=(n_epochs)
+            Predicted target per epoch.
+        
+        """
+        return self.predict(X)
```
Comment (agramfort) [inline mne/decoding/regressor.py]: predict == transform???

File: mne/decoding/regressor.py [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+        
+        """
+        return self.fit(X, y).predict(X)
+    
+    def predict(self, X):
+        """Predict target variable for each trial in X.
+        
+        Parameters
+        ----------
+        X : array, shape=(n_epochs, n_features)
```
Comment (agramfort) [inline mne/decoding/regressor.py]: remove the =

File: mne/decoding/regressor.py [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+            The features for each epochs.
+        
+        Returns
+        -------
+        y_pred : array, shape=(n_epochs, n_targets)
```
Comment (agramfort) [inline mne/decoding/regressor.py]: remove the =

File: mne/decoding/regressor.py [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+            Predicted target variables per epochs.
+        """
+        if not isinstance(X, np.ndarray):
+            raise ValueError("X should be of type ndarray (got %s)."
+                             % type(X))
```
Comment (agramfort) [inline mne/decoding/regressor.py]: remove the check. You have the atleast_2d below

File: mne/decoding/transformer.py [hunk 16]
```diff
@@ -0,0 +1,455 @@
...
+        self.n_jobs = n_jobs
+        self.verbose = verbose
+        self.normalization = normalization
+
+    def fit(self, epochs_data, y):
+        """Compute power spectrum density (PSD) using a multi-taper method
+
+        Parameters
+        ----------
+        epochs_data : array, shape=(n_epochs, n_channels, n_times)
```
Comment (agramfort) [inline mne/decoding/transformer.py]: remove the =

File: mne/decoding/transformer.py [hunk 17]
```diff
@@ -0,0 +1,455 @@
...
+                             % type(epochs_data))
+
+        return self
+
+    def transform(self, epochs_data, y=None):
+        """Compute power spectrum density (PSD) using a multi-taper method
+
+        Parameters
+        ----------
+        epochs_data : array, shape=(n_epochs, n_channels, n_times)
```
Comment (agramfort) [inline mne/decoding/transformer.py]: remove all the = after shape. Just insert an empty space.

File: mne/decoding/transformer.py [hunk 18]
```diff
@@ -0,0 +1,455 @@
...
+                epochs_data = \
+                    band_stop_filter(epochs_data, self.info['sfreq'],
+                                     self.h_freq, self.l_freq,
+                                     filter_length=self.filter_length,
+                                     l_trans_bandwidth=self.h_trans_bandwidth,
+                                     h_trans_bandwidth=self.l_trans_bandwidth,
+                                     method=self.method,
+                                     iir_params=self.iir_params,
+                                     picks=self.picks, n_jobs=self.n_jobs,
+                                     copy=False, verbose=False)
```
Comment (agramfort) [inline mne/decoding/transformer.py]: this block above is a duplicate of what's in the raw.filter method. It should be refactored.

File: mne/viz/decoding.py [hunk 19]
```diff
@@ -11,6 +11,135 @@
...
 
+def plot_patterns(patterns, info, times=None, ch_type=None, layout=None,
+                  vmin=None, vmax=None, cmap='RdBu_r', sensors=True,
+                  colorbar=True, scale=None, scale_time=1e3, unit=None,
+                  res=64, size=1, cbar_fmt='%3.1f',
+                  time_format='%01d ms', proj=False, show=True,
+                  show_names=False, title=None, mask=None,
+                  mask_params=None, outlines='head', contours=6,
+                  image_interp='bilinear', average=None, head_pos=None):
+    """Plot topographic patterns of a linear model
```
Comment (agramfort) [inline mne/viz/decoding.py]: insert empty line

**Comment by agramfort** [comment]
travis is not happy

**Comment by agramfort** [comment]
and there is no test.

@trachelr looks good but please finish PRs before opening new ones.

File: mne/decoding/transformer.py (hunk 18 above)
Comment (trachelr) [inline mne/decoding/transformer.py]: What do you mean by refactored?

File: mne/decoding/classifier.py [hunk 20]
```diff
@@ -1,455 +1,141 @@
...
 
         Parameters
         ----------
-        epochs_data : array, shape=(n_epochs, n_channels, n_times)
-            The data.
-        y : None
-            Not used.
+        X : array, shape (n_epochs, n_features)
+            The data to estimate the coeffiscient.
+        y : array, shape (n_epochs)
```
Comment (agramfort) [inline mne/decoding/classifier.py]: (n_epochs) -> (n_epochs,)

check everywhere

File: examples/decoding/plot_decoding_linear_classifier.py [hunk 21]
```diff
@@ -0,0 +1,84 @@
+"""
+===============================================================
+Decoding sensor space data using a linear classifier
+Patterns and filters visualization of the linear classifier
+===============================================================
```
Comment (kingjr) [inline examples/decoding/plot_decoding_linear_classifier.py]: English syntax

File: examples/decoding/plot_decoding_linear_classifier.py [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+y = np.concatenate(y)
+
+# import a linear classifier from mne.decoding
+from mne.decoding import LinearClassifier
+from sklearn.preprocessing import StandardScaler
+
+clf = LinearClassifier()
+sc = StandardScaler()
+
+X = sc.fit_transform(X)
```
Comment (kingjr) [inline examples/decoding/plot_decoding_linear_classifier.py]: Shouldn't this be in a cross val?

File: mne/decoding/tests/test_regressor.py [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+    # test patterns have been computed
+    assert_true(reg.patterns_ is not None)
+    # test filters have been computed
+    assert_true(reg.filters_ is not None)
+
+    # test predict
+    reg.predict(epochs_data)
+
+    # test classifier without a coef_ attribute
+    reg = LinearRegressor(RandomForestRegressor())
```
Comment (agramfort) [inline mne/decoding/tests/test_regressor.py]: set n_estimators=1 to gain computation time

File: mne/decoding/tests/test_transformer.py [hunk 24]
```diff
@@ -0,0 +1,156 @@
...
+    picks = picks[1:13:3]
+
+    epochs = Epochs(raw, events, event_id, tmin, tmax, picks=picks,
+                    baseline=(None, 0), preload=True)
+    epochs_data = epochs.get_data()
+    scaler = Scaler(epochs.info)
+    y = epochs.events[:, -1]
+
+    # np invalid divide value warnings
+    with warnings.catch_warnings(record=True):
```
Comment (agramfort) [inline mne/decoding/tests/test_transformer.py]: insert :

warnings.simplefilter('always')  # enable b/c these tests throw warnings

here. Not at top of the file.

File: mne/decoding/transformer.py [hunk 25]
```diff
@@ -0,0 +1,524 @@
...
+        - l_freq > h_freq: band-stop filter
+        - l_freq is not None, h_freq is None: low-pass filter
+        - l_freq is None, h_freq is not None: high-pass filter
+
+    If n_jobs > 1, more memory is required as "len(picks) * n_times"
+    additional time points need to be temporarily stored in memory.
+
+    Parameters
+    ----------
+    info : dict
```
Comment (agramfort) [inline mne/decoding/transformer.py]: instance of Info

File: mne/decoding/transformer.py [hunk 26]
```diff
@@ -0,0 +1,524 @@
...
+        self.iir_params = iir_params
+
+    def fit(self, epochs_data, y):
+        """Filters data
+
+        Parameters
+        ----------
+        epochs_data : array, shape (n_epochs, n_channels, n_times)
+            The data.
+        y : array
```
Comment (agramfort) [inline mne/decoding/transformer.py]: shape?

File: mne/decoding/base.py [hunk 27]
```diff
@@ -0,0 +1,174 @@
...
+#
+# License: BSD (3-clause)
+
+import inspect
+import warnings
+import six
+import numpy as np
+
+
+class BaseEstimator(object):
```
Comment (kingjr) [inline mne/decoding/base.py]: Is this copy pasted from sklearn? If so shouldn't we just import it ?

File: mne/decoding/transformer.py [hunk 28]
```diff
@@ -0,0 +1,524 @@
...
+        return self
+
+    def transform(self, epochs_data, y=None):
+        """Filters data
+
+        Parameters
+        ----------
+        epochs_data : array, shape (n_epochs, n_channels, n_times)
+            The data.
+        y : None
```
Comment (agramfort) [inline mne/decoding/transformer.py]: None | ???

File: mne/viz/decoding.py [hunk 29]
```diff
@@ -12,6 +12,138 @@
...
```
Comment (agramfort) [inline mne/viz/decoding.py]: missing empty line.

File: mne/viz/decoding.py [hunk 30]
```diff
@@ -12,6 +12,138 @@
...
+    Parameters
+    ----------
+    patterns : array of floats
```
Comment (agramfort) [inline mne/viz/decoding.py]: shape?

File: mne/decoding/classifier.py [hunk 31]
```diff
@@ -1,455 +1,141 @@
-# Authors: Mainak Jas <mainak@neuro.hut.fi>
+# Authors: Romain Trachel <trachelr@gmail.com>
```
Comment (kingjr) [inline mne/decoding/classifier.py]: Maybe leave mainak?

File: mne/decoding/classifier.py [hunk 32]
```diff
@@ -1,455 +1,141 @@
...
+        computes the attribute patterns_ using [1].
 
         Parameters
         ----------
-        epochs_data : array, shape=(n_epochs, n_channels, n_times)
-            The data.
-        y : None
-            Not used.
+        X : array, shape (n_epochs, n_features)
+            The data to estimate the coeffiscient.
```
Comment (kingjr) [inline mne/decoding/classifier.py]: Typo

File: mne/decoding/classifier.py [hunk 33]
```diff
@@ -1,455 +1,141 @@
...
-        self.info = info
-
-    def fit(self, epochs_data, y):
-        """Concatenate different channels into a single feature vector
+        self : instance of LinearClassifier
+            Returns the modified instance.
 
-        Parameters
+        References
         ----------
```
Comment (kingjr) [inline mne/decoding/classifier.py]: Missing ref

File: mne/decoding/classifier.py [hunk 34]
```diff
@@ -1,455 +1,141 @@
...
-            raise ValueError("epochs_data should be of type ndarray (got %s)."
-                             % type(epochs_data))
-
-        return self
+        return self.fit(X, y).transform(X)
 
-    def transform(self, epochs_data, y=None):
-        """Compute power spectrum density (PSD) using a multi-taper method
+    def predict(self, X):
+        """Predict class labels for each trial in X.
```
Comment (kingjr) [inline mne/decoding/classifier.py]: I think we don't use 'trial' in mne or sklearn . Probably better is 'epoch'

File: mne/decoding/regressor.py [hunk 35]
```diff
@@ -0,0 +1,104 @@
...
+    Attributes
+    ----------
+    filters_ : ndarray
+        If fit, the filters used to decompose the data, else None.
+    patterns_ : ndarray
+        If fit, the patterns used to restore M/EEG signals, else None.
+
+    References
+    ----------
+    """
```
Comment (kingjr) [inline mne/decoding/regressor.py]: Missing ref

File: mne/decoding/regressor.py [hunk 36]
```diff
@@ -0,0 +1,104 @@
+# Authors: Romain Trachel <trachelr@gmail.com>
+#          Alexandre Gramfort <alexandre.gramfort@telecom-paristech.fr>
+#
+# License: BSD (3-clause)
+
+import numpy as np
+from .base import BaseEstimator
+
+
+class LinearRegressor(BaseEstimator):
```
Comment (kingjr) [inline mne/decoding/regressor.py]: The codes between classifiers and regressors are identical aren't they? If so just have a linear estimator object that checks whether the values in y are integers.

**Comment by kingjr** [comment]
I m just having a quick look on my phone so I may have missed some things but I m unclear on the 'filter' option and attribute. Can you detail its purple ?

File: examples/decoding/plot_decoding_linear_classifier.py [hunk 37]
```diff
@@ -0,0 +1,84 @@
...
+# Decoding in sensor space using a linear classifier
+
+# Make arrays X and y such that :
+# X is 2d with X.shape[0] is the total number of epochs to classify
+# y is filled with integers coding for the class to predict
+# We must have X.shape[0] equal to y.shape[0]
+X = [e.get_data()[:, :, 0] for e in epochs_list]
+y = [k * np.ones(len(this_X)) for k, this_X in enumerate(X)]
+X = np.concatenate(X)
+y = np.concatenate(y)
```
Comment (dengemann) [inline examples/decoding/plot_decoding_linear_classifier.py]: And this should be in a an Epochs Vectorizer #2313

Comment (mainakjas) [inline examples/decoding/plot_decoding_linear_classifier.py]: yes except we need a fix in sklearn to make it work :)

File: mne/decoding/base.py (hunk 27 above)
Comment (trachelr) [inline mne/decoding/base.py]: Yep it is a copy-paste from sklearn... @agramfort told me to do this.
I think he wanted to avoid dependencies for base classes

Comment (agramfort) [inline mne/decoding/base.py]: yes we don't want a strong dependency on sklearn.

**Comment by trachelr** [comment]
Hey guys
That's pretty much all for this PR.
Could you give me one more review before merging please?
There's other PR to open, such as refactor code in CSP (_regularized_covariance), FilterEstimator, rename ConcatenateChannels etc...

File: doc/source/whats_new.rst [hunk 38]
```diff
@@ -11,6 +11,10 @@ Current
 Changelog
 ~~~~~~~~~
 
+    - Add support for plotting patterns/filters in :class:`mne.decoding.csp.CSP` and :class:`mne.decoding.base.LinearModel` by `Romain Trachel`_
```
Comment (mainakjas) [inline doc/source/whats_new.rst]: you should add this at the bottom

File: examples/decoding/plot_linear_model_patterns.py [hunk 39]
```diff
@@ -0,0 +1,78 @@
...
+meg_epochs = epochs.pick_types(meg=True, eeg=False, copy=True)
+meg_data = meg_epochs.get_data().reshape(len(labels), -1)
+eeg_epochs = epochs.pick_types(meg=False, eeg=True, copy=True)
+eeg_data = eeg_epochs.get_data().reshape(len(labels), -1)
+
+###############################################################################
+# Decoding in sensor space using a linear classifier
+
+clf = LogisticRegression()
+sc = StandardScaler()
```
Comment (mainakjas) [inline examples/decoding/plot_linear_model_patterns.py]: out of curiosity, what happens when you use a `MinMaxScaler`? I think @alexandrebarachant used that in his pipeline before the LR model

File: mne/decoding/base.py [hunk 40]
```diff
@@ -0,0 +1,602 @@
...
+
+    def fit(self, X, y):
+        """Estimate the coeffiscient of the linear model.
+        Save the coeffiscient in the attribute filters_ and
+        computes the attribute patterns_ using [1].
+
+        Parameters
+        ----------
+        X : array, shape (n_epochs, n_features)
+            The data to estimate the coeffiscient.
```
Comment (mainakjas) [inline mne/decoding/base.py]: coefficient

File: mne/decoding/base.py [hunk 41]
```diff
@@ -0,0 +1,602 @@
...
+            'scale' (tuple) for what the center and scale of the head
+            should be relative to the electrode locations.
+
+        Returns
+        -------
+        fig : instance of matplotlib.figure.Figure
+           The figure.
+        """
+
+        from .. import EvokedArray
```
Comment (mainakjas) [inline mne/decoding/base.py]: why nested?

File: mne/decoding/base.py [hunk 42]
```diff
@@ -0,0 +1,602 @@
...
```
Comment (mainakjas) [inline mne/decoding/base.py]: same question here

File: mne/decoding/base.py [hunk 43]
```diff
@@ -0,0 +1,602 @@
...
+    def plot_patterns(self, info, times=None, ch_type=None, layout=None,
+                      vmin=None, vmax=None, cmap='RdBu_r', sensors=True,
+                      colorbar=True, scale=None, scale_time=1e3, unit='a.u.',
+                      res=64, size=1, cbar_fmt='%3.1f',
+                      name_format='%01d ms', proj=False, show=True,
+                      show_names=False, title=None, mask=None,
+                      mask_params=None, outlines='head', contours=6,
+                      image_interp='bilinear', average=None, head_pos=None):
+        """
+        Plot topographic patterns of the linear model
```
Comment (mainakjas) [inline mne/decoding/base.py]: maybe you should explain again here what a "topographic pattern" is

File: mne/decoding/base.py [hunk 44]
```diff
@@ -0,0 +1,602 @@
...
+    def plot_filters(self, info, times=None, ch_type=None, layout=None,
+                     vmin=None, vmax=None, cmap='RdBu_r', sensors=True,
+                     colorbar=True, scale=None, scale_time=1e3, unit='a.u.',
+                     res=64, size=1, cbar_fmt='%3.1f',
+                     name_format='%01d ms', proj=False, show=True,
+                     show_names=False, title=None, mask=None,
+                     mask_params=None, outlines='head', contours=6,
+                     image_interp='bilinear', average=None, head_pos=None):
+        """
+        Plot topographic filters of the linear model
```
Comment (mainakjas) [inline mne/decoding/base.py]: same comment here. Explain again what a filter is.

File: mne/decoding/transformer.py [hunk 45]
```diff
@@ -30,23 +31,28 @@ class Scaler(TransformerMixin):
 
     Attributes
     ----------
-    `scale_` : dict
+    info : instance of Info
+        The measurement info
+    ch_mean_ : dict
         The mean value for each channel type
-    `ch_std_` : array
+    std_ : dict
```
Comment (mainakjas) [inline mne/decoding/transformer.py]: this looks a bit brutal to me. Don't we need deprecation?

File: mne/decoding/transformer.py [hunk 46]
```diff
@@ -86,17 +91,17 @@ def transform(self, epochs_data, y=None):
...
-        y : None
-            Not used.
+        y : None | array, shape (n_epochs,)
+            The label for each epoch.
+            If None not used. Defaults to None.
 
         Returns
         -------
-        X : array of shape (n_epochs, n_channels * n_times)
+        X : array, shape (n_epochs, n_channels, n_times)
```
Comment (mainakjas) [inline mne/decoding/transformer.py]: why did you change this? You get a 2D array not 3D

**Comment by mainakjas** [comment]
Code looks pretty clean to me @trachelr !

File: mne/decoding/transformer.py (hunk 45 above)
Comment (trachelr) [inline mne/decoding/transformer.py]: I just changed the doc here, because the attribute is std_
We discussed about this class during the sprint, attributes names are messed up.

File: examples/decoding/plot_linear_model_patterns.py (hunk 39 above)
Comment (trachelr) [inline examples/decoding/plot_linear_model_patterns.py]: It change the range of patterns/filters values, not the their topography...

File: mne/decoding/base.py [hunk 47]
```diff
@@ -0,0 +1,602 @@
...
+        score : float
+            Score of the linear model
+
+        """
+        return self.model.score(X, y)
+
+    def plot_patterns(self, info, times=None, ch_type=None, layout=None,
+                      vmin=None, vmax=None, cmap='RdBu_r', sensors=True,
+                      colorbar=True, scale=None, scale_time=1e3, unit='a.u.',
+                      res=64, size=1, cbar_fmt='%3.1f',
```
Comment (kingjr) [inline mne/decoding/base.py]: @trachelr @agramfort I think that all patterns can (and, in my opinion *should*) be inverted back to tesla or volt. You just have to invert_transform the scaler.

Comment (trachelr) [inline mne/decoding/base.py]: I added an inverse_transform to the scaler (and to any class in mne.decoding.transformer)
but there's an issue because it reshapes (n_epochs, n_channels, n_times) and patterns are shape (n_channes, n_times). 
I think reshaping with (-1, n_channels, n_times) should fix this issue... I'll try it later and commit if it works.

**Comment by trachelr** [comment]
@agramfort, could you tell me how to solve this non-mergable issue please?

**Comment by agramfort** [comment]
try to rebase but if it gets messy don't bother and just make travis happy. I'll take care of the merge.

**Comment by trachelr** [comment]
Rebase done, the code looks clean... 
Let see if travis is happy (I think he should)
Ready for merging if it ok for you @agramfort 
Next time, I'll make shorter PR...

File: mne/decoding/tests/test_csp.py [hunk 48]
```diff
@@ -93,3 +93,43 @@ def test_regularized_csp():
...
+    epochs = Epochs(raw, events, event_id, tmin, tmax, picks=picks,
+                    baseline=(None, 0), preload=True)
+    epochs_data = epochs.get_data()
+
+    n_components = 3
+    csp = CSP(n_components=n_components, reg='lws')
+
+    csp.fit(epochs_data, epochs.events[:, -1])
+    csp.plot_filters(epochs.info)
+    csp.plot_filters(epochs.info, components=np.arange(10))
```
Comment (agramfort) [inline mne/decoding/tests/test_csp.py]: how long does the tests take to run for you? they take ages for some reason on my machine :(

Comment (agramfort) [inline mne/decoding/tests/test_csp.py]: tests gets stalled on my box when running them... :(

Comment (trachelr) [inline mne/decoding/tests/test_csp.py]: it runs pretty quickly... 
line135 have been removed from the test and reg changed to .9
Maybe it is due to some convergence issues in lws

**Comment by agramfort** [comment]
this is really weird. When running the test from ipython is runs but with

nosetests mne/decoding/tests/test_csp.py

it stalls. Any idea? Can you reproduce?

**Comment by trachelr** [comment]
Nop I didnt reproduce it. I ran tests in 9.712s with nosetests.
The problem must come from test_plot_patterns/test_plot_filters.
Can you try it with show=False (last commit) to see if the problem comes from matplotlib.

**Comment by agramfort** [comment]
yes setting show=False fixes the problem.

**Comment by agramfort** [comment]
merged manually after a tiny commit:

1494d92255f48e0b1d1c782684dfd123a9dd034d

thanks @trachelr !

**Comment by agramfort** [comment]
@trachelr open a new PR for the remaining parts or just open an issue so we don't forget.
